File: csharp_dotnetcore/19.Bot-Authentication/SimpleGraphClient.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using Microsoft.Graph;

namespace Bot_Authentication
{
    // This class is a wrapper for the Microsoft Graph API
    // See: https://developer.microsoft.com/en-us/graph
    public class SimpleGraphClient
    {
        private readonly string _token;

        public SimpleGraphClient(string token)
        {
            this._token = token;
        }

        // Sends an email on the users behlaf using the Microsoft Graph API
        public async Task<bool> SendMailAsync(string toAddress, string subject, string content)
        {
            try
            {
                var graphClient = this.GetAuthenticatedClient();

                var recipients = new List<Recipient>
                {
                    new Recipient
                    {
                        EmailAddress = new EmailAddress
                        {
                            Address = toAddress,
                        },
                    },
                };

                // Create the message.
                var email = new Message
                {
                    Body = new ItemBody
                    {
                        Content = content,
                        ContentType = BodyType.Text,
                    },
                    Subject = subject,
                    ToRecipients = recipients,
                };

                // Send the message.
                await graphClient.Me.SendMail(email, true).Request().PostAsync();
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }

        // Gets recent unread mail for the user using the Microsoft Graph API
        public async Task<Message[]> GetRecentUnreadMailAsync()
        {
            var graphClient = this.GetAuthenticatedClient();
            var messages =
                await graphClient.Me.MailFolders.Inbox.Messages.Request().GetAsync();
            var from = DateTime.Now.Subtract(TimeSpan.FromMinutes(60));
            var unreadMessages = new List<Message>();

            var done = false;
            while (messages?.Count > 0 && !done)
            {
                foreach (var message in messages)
                {
                    if (message.ReceivedDateTime.HasValue && message.ReceivedDateTime.Value >= from)
                    {
                        if (message.IsRead.HasValue && !message.IsRead.Value)
                        {
                            unreadMessages.Add(message);
                            if (unreadMessages.Count >= 5)
                            {
                                done = true;
                            }
                        }
                    }
                    else
                    {
                        done = true;
                    }
                }

                messages = await messages.NextPageRequest.GetAsync();
            }

            return unreadMessages.ToArray();
        }

        // Get information about the user.
        public async Task<User> GetMeAsync()
        {
            var graphClient = this.GetAuthenticatedClient();
            var me = await graphClient.Me.Request().GetAsync();
            return me;
        }

        // gets information about the user's manager
        public async Task<User> GetManagerAsync()
        {
            var graphClient = this.GetAuthenticatedClient();
            var manager = await graphClient.Me.Manager.Request().GetAsync() as User;
            return manager;
        }

        // Gets the user's photo
        public async Task<PhotoResponse> GetPhotoAsync()
        {
            var photoResponse =
                await new HttpClient().GetStreamWithAuthAsync(
                    this._token,
                    "https://graph.microsoft.com/v1.0/me/photo/$value");
            if (photoResponse != null)
            {
                photoResponse.Base64String = $"data:{photoResponse.ContentType};base64," +
                                             Convert.ToBase64String(photoResponse.Bytes);
            }

            return photoResponse;
        }

        // Get an Authenticated Microsoft Graph client using the token issued to the user.
        private GraphServiceClient GetAuthenticatedClient()
        {
            var graphClient = new GraphServiceClient(
                new DelegateAuthenticationProvider(
                    requestMessage =>
                    {
                        // Append the access token to the request.
                        requestMessage.Headers.Authorization = new AuthenticationHeaderValue("bearer", _token);

                        // Get event times in the current time zone.
                        requestMessage.Headers.Add("Prefer", "outlook.timezone=\"" + TimeZoneInfo.Local.Id + "\"");

                        return Task.CompletedTask;
                    }));
            return graphClient;
        }
    }
}

File: samples/javascript_nodejs/24.bot-authentication-msgraph/simple-graph-client.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { Client } = require('@microsoft/microsoft-graph-client');

/**
 * This class is a wrapper for the Microsoft Graph API.
 * See: https://developer.microsoft.com/en-us/graph for more information.
 */
class SimpleGraphClient {
    constructor(token) {
        if (!token || !token.trim()) {
            throw new Error('SimpleGraphClient: Invalid token received.');
        }

        this._token = token;

        // Get an Authenticated Microsoft Graph client using the token issued to the user.
        this.graphClient = Client.init({
            authProvider: (done) => {
                done(null, this._token); // First parameter takes an error if you can't get an access token.
            }
        });
    }

    /**
     * Sends an email on the user's behalf.
     * @param {string} toAddress Email address of the email's recipient.
     * @param {string} subject Subject of the email to be sent to the recipient.
     * @param {string} content Email message to be sent to the recipient.
     */
    async sendMail(toAddress, subject, content) {
        if (!toAddress || !toAddress.trim()) {
            throw new Error('SimpleGraphClient.sendMail(): Invalid `toAddress` parameter received.');
        }
        if (!subject || !subject.trim()) {
            throw new Error('SimpleGraphClient.sendMail(): Invalid `subject`  parameter received.');
        }
        if (!content || !content.trim()) {
            throw new Error('SimpleGraphClient.sendMail(): Invalid `content` parameter received.');
        }

        // Create the email.
        const mail = {
            body: {
                content: content, // `Hi there! I had this message sent from a bot. - Your friend, ${ graphData.displayName }!`,
                contentType: 'Text'
            },
            subject: subject, // `Message from a bot!`,
            toRecipients: [{
                emailAddress: {
                    address: toAddress
                }
            }]
        };

        // Send the message.
        return await this.graphClient
            .api('/me/sendMail')
            .post({ message: mail }, (error, res) => {
                if (error) {
                    throw error;
                } else {
                    return res;
                }
            });
    }

    /**
     * Gets recent mail the user has received within the last hour and displays up to 5 of the emails in the bot.
     */
    async getRecentMail() {
        return await this.graphClient
            .api('/me/messages')
            .version('beta')
            .top(5)
            .get().then((res) => {
                return res;
            });
    }

    /**
     * Collects information about the user in the bot.
     */
    async getMe() {
        return await this.graphClient
            .api('/me')
            .get().then((res) => {
                return res;
            });
    }

    /**
     * Collects the user's manager in the bot.
     */
    async getManager() {
        return await this.graphClient
            .api('/me/manager')
            .version('beta')
            .select('displayName')
            .get().then((res) => {
                return res;
            });
    }
}

exports.SimpleGraphClient = SimpleGraphClient;
